# Курсовая работа по дисциплине Структуры и алгоритмы обработки данных


## Выполнил студент группы БВТ2005
## Куликов Павел Максимович

# Задание:

Необходимо найти решение нескольких задач в соответствии с вариантом. 

Каждая задача имеет стоимость в 2 балла. Итоговая оценка выставляется в соответствии с таблицей:

Кол-во баллов | Оценка
:-------- |:-----
0-10 | Неудовлетворительно
12-14 | Удовлетворительно
16-18| Хорошо
20 | Отлично

## Вариант 8

### Задача 1
Юный тренер покемонов Сергей Б. обнаружил большой дом, состоящий из n квартир, выстроенных в ряд слева направо, причем в каждую квартиру можно зайти с улицы, из каждой квартиры можно выйти на улицу, а также из каждой квартиры можно зайти в соседнюю слева или справа квартиру. Из квартиры номер 1 можно попасть только в квартиру номер 2, а из квартиры номер n можно попасть только в квартиру номер n - 1.

Так сложилось, что в каждой из этих квартир находится ровно один покемон какого-то типа. Возможно, что в каких-то квартирах находятся покемоны одинаковых типов. Сергей Б. попросил жителей дома пустить его к ним в квартиры, чтобы поймать покемонов. Посовещавшись, жители дома решили, что разрешат Сергею Б. зайти с улицы ровно в одну квартиру, посетить несколько квартир, а затем выйти из какой-то квартиры на улицу, причем они не разрешили Сергею Б. посещать никакую квартиру более одного раза.

Сергей Б. очень обрадовался, но как человек интеллигентный, решил посетить как можно меньше квартир, но при этом, естественно, собрать покемонов всех типов, которые есть в доме (иначе зачем тогда это всё). Перед вами стоит задача помочь ему и определить это минимальное число квартир.

Входные данные\
В первой строке следует целое число n (1 ≤ n ≤ 100 000) — количество квартир в доме.

Во второй строке следует строка s длины n, состоящая из строчных и прописных букв английского алфавита, причем i-я буква равна типу покемона, который находится в квартире номер i.

Выходные данные\
Выведите минимальное число квартир, которые должен посетить Сергей Б., чтобы поймать покемонов всех типов, которые есть в доме.

Примеры\
входные данные\
3\
AaA\
выходные данные\
2\
входные данные\
7\
bcAAcbc\
выходные данные\
3\
входные данные\
6\
aaBCCe\
выходные данные\
5\
Примечание\
В первом тестовом примере Сергей Б. может, например, начать с квартиры 1 и закончить в квартире 2.

Во втором тестовом примере Сергей Б. может, например, начать с квартиры 4 и закончить в квартире 6.

В третьем тестовом примере Сергей Б. должен начать с квартиры 2 и закончить в квартире 6.

In [14]:
import math
import random

def solution(n):
    pokedex = []
    string = ""
    len = [-1] * n
    for i in range(n):
        pokemon = random.randint(0,1)
        if pokemon:
            string += str(chr(random.randint(65,90)))
        else:
            string += str(chr(random.randint(97, 122)))

    for i in string:
        if i not in pokedex:
            pokedex.append(i)

    for i in pokedex:
        last = -1
        for j in range(n):
            if string[j] == i:
                last = j
            if last == -1:
                len[j] = math.inf
            else:
                len[j] = max(len[j], j - last + 1)

    return min(len)

print("Количество квартир: ", end="")
n = int(input())
print("Ответ:",solution(n))

Количество квартир: 100000
Ответ: 100


### Задача 2
Массив 𝑎1,𝑎2,…,𝑎𝑛 называется хорошим, если и только если для любого подотрезка 1≤𝑙≤𝑟≤𝑛, выполняется следующее: 𝑎𝑙+𝑎𝑙+1+…+𝑎𝑟=1/2(𝑎𝑙+𝑎𝑟)⋅(𝑟−𝑙+1).

Вам дан массив целых чисел 𝑎1,𝑎2,…,𝑎𝑛. За одну операцию вы можете заменить один любой элемент массива на любое вещественное число. Определите минимальное число операций, необходимое, чтобы сделать массив хорошим.

Входные данные\
Первая строка содержит одно целое число 𝑡 (1≤𝑡≤100): количество наборов входных данных.

Далее следуют описания 𝑡 наборов входных данных, по две строки на набор.

Первая строка содержит одно целое число 𝑛 (1≤𝑛≤70): количество чисел в массиве.

Вторая строка содержит 𝑛 целых чисел 𝑎1,𝑎2,…,𝑎𝑛 (−100≤𝑎𝑖≤100): изначальный массив.

Выходные данные\
Для каждого набора входных данных выведите одно целое число: минимальное число элементов, которое нужно заменить, чтобы массив стал хорошим.

Пример\
входные данные
5\
4\
1 2 3 4\
4\
1 1 2 2\
2\
0 -1\
6\
3 -2 4 -1 -4 0\
1\
-100\
выходные данные\
0\
2\
0\
3\
0\
Примечание\
В первом примере массив хороший изначально.

Во втором примере один из возможных хороших массивов — [1,1,1⎯⎯,1⎯⎯] (замененные элементы подчеркнуты).

В третьем примере массив хороший изначальное.

Во четвертом примере один из возможных хороших массивов — [−2.5⎯⎯⎯⎯⎯⎯⎯⎯,−2,−1.5⎯⎯⎯⎯⎯⎯⎯⎯,−1,−0.5⎯⎯⎯⎯⎯⎯⎯⎯,0].

In [8]:
import random


def solution(arr):
    if len(arr) == 1:
        return 0
    diference = {}
    res = 0
    for i in range(len(arr)):
        for j in range(i+1, len(arr)):
            dif = (arr[j] - arr[i])/(j - i)
            if dif not in diference.keys():
                diference[dif] = [1, j]
            else:
                diference[dif][0] += 1
                diference[dif][1] = j
    max_dif = sorted(diference.items(), key= lambda x: x[1][0]).pop()[0]
    max_dif_index = sorted(diference.items(), key= lambda x: x[1][0]).pop()[1][1]
    for i in range(len(arr)):
        if arr[i] != arr[max_dif_index] - max_dif * (max_dif_index - i):
            res +=1
    return res

print("Ввод t (1<=t<=100):", end="")
t = int(input())

for i in range(t):
    arr = []
    n = random.randint(1,70)
    for j in range(n):
        arr.append(random.randint(-100,100))
    print(n)
    print(arr)
    print("Ответ: ", solution(arr))
#Проверки алгоритма:
'''arr = [1, 2, 3, 4]
print(solution(arr))'''
'''arr = [1, 1, 2, 2]
print(solution(arr))'''
'''arr = [0, -1]
print(solution(arr))'''
'''arr = [3, -2, 4, -1, -4, 0, 2, 1]
print(solution(arr))'''
'''arr = [-100]
print(solution(arr))'''

Ввод t (1<=t<=100):4
2
[-36, -80]
Ответ:  0
57
[49, -67, 69, -80, 60, -30, -24, -59, -71, 84, 35, -3, -100, -62, -53, 81, -50, -14, 26, 60, -94, 14, -8, 9, -93, 100, -41, 60, -29, 16, -74, -80, 34, 58, -27, 56, -47, 56, -8, 22, 1, -52, -53, 82, 31, -27, -51, 86, 17, 62, 27, 23, -71, -62, 87, -6, 94]
Ответ:  55
23
[-76, 62, -77, -48, 93, 49, -33, -52, -3, -59, -28, -90, 74, 30, -59, 81, -52, -8, 15, -34, -6, -43, -48]
Ответ:  21
18
[12, -91, -36, 94, -35, -84, -90, -61, -33, 78, -23, -82, -99, -92, -36, 10, -69, 2]
Ответ:  15


'arr = [-100]\nprint(solution(arr))'

### Задача 3
Чтобы удовлетворить свою любовь к собиранию носков в пары, Феникс принес свои 𝑛 носков (𝑛 — четное) в магазин по продаже носков. Каждый его носок определенного цвета 𝑐𝑖 и либо левый, либо правый.

Феникс может заплатить один доллар магазину, чтобы:

либо перекрасить какой-то носок в любой цвет 𝑐′ (1≤𝑐′≤𝑛),
либо превратить левый носок в правый,
либо превратить правый носок в левый.
Магазин может производить любое из изменений любое количество раз. Заметим, что цвет левого носка не меняется, когда он превращается в правый, и наоборот.

Два носка образуют пару, если это левый и правый носок одинакового цвета. Какую минимальное количество денег нужно заплатить Фениксу, чтобы собрать 𝑛/2 пар? Каждый носок должен попасть ровно в одну пару.

Входные данные\
Входные данные состоят из нескольких наборов. В первой строке задано одно целое число 𝑡 (1≤𝑡≤1000) — количество наборов входных данных.

В первой строке каждого набора заданы три целых числа 𝑛, 𝑙 и 𝑟 (2≤𝑛≤2⋅105; 𝑛 — четное; 0≤𝑙,𝑟≤𝑛; 𝑙+𝑟=𝑛) — общее количество носков и количество левых и правых носков соответственно.

В следующей строке заданы 𝑛 целых чисел 𝑐𝑖 (1≤𝑐𝑖≤𝑛) — цвета носков. Первые 𝑙 носков — левые, остальные 𝑟 носков — правые.

Гарантируется, что сумма 𝑛 по всем наборам входных данных не превосходит 2⋅105.

Выходные данные\
Для каждого набора входных данных, выведите одно число — минимальное количество денег, которое нужно заплатить Фениксу, чтобы собрать 𝑛/2 пар носков. Каждый носок должен попасть ровно в одну пару.

Пример\
входные данные\
4\
6 3 3\
1 2 3 2 2 2\
6 2 4\
1 1 2 2 2 2\
6 5 1\
6 5 4 3 2 1\
4 0 4\
4 4 4 3\
выходные данные\
2\
3\
5\
3\
Примечание\
В первом наборе, Феникс может заплатить 2 доллара для того, чтобы:

перекрасить носок 1 в цвет 2,\
перекрасить носок 3 в цвет 2.\
В результате он получит 3 пары. Например, пары (1,4), (2,5) и (3,6).\
Во втором наборе, Феникс может заплатить 3 доллара, чтобы:

превратить носок 6 из правого в левый,\
перекрасить носок 3 в цвет 1,\
перекрасить носок 4 в цвет 1.\
В результате он получит 3 пары. Например, пары (1,3), (2,4) и (5,6).

In [1]:
import random


def solution(N, L, R, C):
    cost = 0
    lcnt = []
    rcnt = []
    counter = 0
    for i in range(L):
        lcnt.append(C.pop(i - counter))
        counter += 1
    counter = 0
    for i in range(R):
        rcnt.append(C.pop(i - counter))
        counter += 1
    if len(lcnt) > len(rcnt):
        lcnt, rcnt = rcnt, lcnt
    counter = 0
    for i in range(len(lcnt)):
        if lcnt[i-counter] in rcnt:
            rcnt.pop(rcnt.index(lcnt[i - counter]))
            lcnt.pop(i - counter)
            counter += 1
        else:
            lcnt.pop(i - counter)
            rcnt.pop(i - counter)
            cost += 1
            counter += 1
    if rcnt:
        counter = 0
        for i in range(len(rcnt) // 2):
            lcnt.append(rcnt.pop(i - counter))
            cost += 1
            counter += 1
        counter = 0
        for i in range(len(lcnt)):
            if lcnt[i-counter] in rcnt:
                rcnt.pop(rcnt.index(lcnt[i - counter]))
                lcnt.pop(i - counter)
                counter += 1
            else:
                lcnt.pop(i - counter)
                rcnt.pop(i - counter)
                cost += 1
                counter += 1
    return cost


print("Ввод t (1<=t<=1000):", end="")
t = int(input())

for i in range(t):
    C = []
    N = random.randint(1,10**5)*2
    L = random.randint(0,N)
    R = N - L
    for j in range(N):
        C.append(random.randint(1,N))
    print("N =", N, "L =", L, "R =", R)
    print(C)
    print("Ответ: ", solution(N, L, R, C))
#Проверки алгоритма:
'''N, L, R = 6, 3, 3
C = [1, 2, 3, 2, 2, 2]
print(solution(N, L, R, C))'''
'''N, L, R = 6, 2, 4
C = [1, 1, 2, 2, 2, 2]
print(solution(N, L, R, C))'''
'''N, L, R = 6, 5, 1
C = [6, 5, 4, 3, 2, 1]
print(solution(N, L, R, C))'''
'''N, L, R = 4, 0, 4
C = [4, 4, 4, 3]
print(solution(N, L, R, C))'''

Ввод t (1<=t<=1000):4
N = 20 L = 12 R = 8
[6, 17, 15, 13, 16, 11, 4, 10, 9, 11, 2, 3, 5, 13, 9, 12, 10, 3, 20, 1]
Ответ:  7
N = 14 L = 3 R = 11
[2, 5, 3, 14, 7, 6, 4, 2, 5, 2, 9, 5, 7, 12]
Ответ:  8
N = 8 L = 6 R = 2
[4, 3, 7, 4, 7, 1, 8, 4]
Ответ:  5
N = 2 L = 1 R = 1
[1, 2]
Ответ:  1


'N, L, R = 4, 0, 4\nC = [4, 4, 4, 3]\nprint(solution(N, L, R, C))'

### Задача 4
У Монокарпа есть две строки 𝑠 и 𝑡 одинаковой длины, состоящие из латинских букв «a» и «b».

Монокарп хочет сделать строки 𝑠 и 𝑡 одинаковыми. Для этого он может выполнять следующие операции: выбрать в строке 𝑠 позицию 𝑝𝑜𝑠1, выбрать в строке 𝑡 позицию 𝑝𝑜𝑠2 и поменять местами буквы 𝑠𝑝𝑜𝑠1 и 𝑡𝑝𝑜𝑠2.

Перед вами стоит задача определить минимальное количество операций, которые должен сделать Монокарп, чтобы сделать строки 𝑠 и 𝑡 одинаковыми, а также вывести сами операции. Если невозможно сделать строки одинаковыми, сообщите об этом.

Входные данные\
В первой строке следует целое число 𝑛 (1≤𝑛≤2⋅105) — длина строк 𝑠 и 𝑡.

Во второй строке следует строка 𝑠 длины 𝑛, состоящая из латинских букв «a» и «b».

В третьей строке следует строка 𝑡 длины 𝑛, состоящая из латинских букв «a» и «b».

Выходные данные\
Если невозможно сделать строки одинаковыми, выведите −1.

В противном случае, в первую строку выведите целое число 𝑘 — минимальное количество операций, которые нужно сделать, чтобы сделать строки одинаковыми. В каждой из следующих 𝑘 строк выведите по два целых числа — позицию в строке 𝑠 и позицию в строке 𝑡, которые нужно использовать для обмена букв во время очередной операции.

Примеры\
входные данные\
4\
abab\
aabb\
выходные данные\
2\
3 3\
3 2\
входные данные\
1\
a\
b\
выходные данные\
-1\
входные данные\
8\
babbaabb\
abababaa\
выходные данные\
3\
2 6\
1 3\
7 8\
Примечание\
В первом примере достаточно двух операций обмена. Например, можно сначала поменять местами третью букву в строке 𝑠 с третьей буквой в строке 𝑡. После этого 𝑠= «abbb», 𝑡= «aaab». Затем нужно поменять третью букву в строке 𝑠 со второй буквой в строке 𝑡. После этого обе строки 𝑠 и 𝑡 будут равны «abab».

Во втором примере невозможно сделать строки одинаковыми.

In [13]:
import random


def solution(n, s, t):
    pos01 = []
    pos10 = []
    pos_chenge = []
    res = 0
    for i in range(n):
        if s[i] != t[i]:
            if s[i] == 'a':
                pos01.append(i)
            else:
                pos10.append(i)
    if len(pos01) % 2 != len(pos10) % 2:
        res = -1
        print(res)
        return
    if len(pos01) % 2 == 0 and len(pos10) % 2 == 0:
        res = (len(pos01)/2)+(len(pos10)/2)
    else:
        res = ((len(pos01) - 1)/2)+((len(pos10) - 1)/2) + 2
    print(int(res))
    if len(pos01) > 1:
        for i in range(0, len(pos01), 2):
            pos_chenge.append([pos01[i]+1, pos01[i + 1]+1])
    if len(pos10) > 1:
        for i in range(0, len(pos10), 2):
            pos_chenge.append([pos10[i]+1, pos10[i + 1]+1])
    if len(pos01) % 2:
        x = len(pos01)-1
        y = len(pos01)-1
        pos_chenge.append([pos01[x]+1, pos01[x]+1])
        pos_chenge.append([pos01[x]+1, pos10[y]+1])
    for i in pos_chenge:
        print(i)


n = random.randint(0, 2*(10**5))
s = ""
t = ""
for i in range(n):
    s += ("a" if random.randint(0, 1) else "b")
    t += ("a" if random.randint(0, 1) else "b")
solution(n, s, t)
#Проверки алгоритма:
'''n = 4
s = "abab"
t = "aabb"
solution(n, s, t)'''
'''n = 1
s = "a"
t = "b"
solution(n, s, t)'''
'''n = 8
s = "babbaabb"
t = "abababaa"
solution(n, s, t)'''

17184
[2, 10]
[11, 17]
[25, 26]
[31, 33]
[36, 39]
[41, 43]
[49, 57]
[60, 61]
[63, 66]
[72, 77]
[90, 95]
[104, 105]
[106, 110]
[124, 126]
[127, 135]
[136, 139]
[142, 145]
[148, 149]
[155, 159]
[161, 165]
[166, 174]
[176, 177]
[178, 181]
[182, 183]
[186, 188]
[191, 199]
[204, 205]
[212, 217]
[220, 221]
[223, 226]
[230, 231]
[235, 241]
[243, 244]
[245, 250]
[251, 252]
[254, 257]
[259, 267]
[276, 278]
[284, 285]
[287, 288]
[293, 296]
[300, 302]
[313, 315]
[318, 319]
[327, 330]
[335, 338]
[341, 343]
[345, 349]
[351, 352]
[354, 355]
[360, 365]
[368, 369]
[372, 378]
[384, 386]
[388, 393]
[397, 403]
[405, 411]
[415, 417]
[420, 423]
[426, 434]
[435, 441]
[445, 448]
[450, 452]
[461, 465]
[481, 482]
[485, 493]
[497, 498]
[501, 516]
[521, 528]
[530, 535]
[537, 538]
[542, 545]
[548, 550]
[558, 562]
[565, 566]
[577, 584]
[586, 587]
[598, 601]
[605, 609]
[610, 614]
[618, 626]
[630, 634]
[636, 639]
[640, 642]
[647, 649]
[650, 653]
[655, 659]
[670, 685]
[688, 689]
[705, 707]
[710, 713]
[717, 718]
[719,

[11173, 11176]
[11178, 11183]
[11185, 11187]
[11188, 11190]
[11195, 11197]
[11201, 11202]
[11203, 11209]
[11214, 11219]
[11220, 11226]
[11233, 11238]
[11242, 11243]
[11245, 11253]
[11261, 11265]
[11266, 11267]
[11269, 11271]
[11272, 11291]
[11292, 11293]
[11295, 11296]
[11301, 11306]
[11307, 11310]
[11311, 11316]
[11317, 11319]
[11320, 11325]
[11328, 11334]
[11335, 11337]
[11339, 11345]
[11347, 11349]
[11350, 11352]
[11356, 11357]
[11358, 11359]
[11364, 11370]
[11373, 11377]
[11380, 11384]
[11388, 11395]
[11399, 11401]
[11404, 11412]
[11415, 11430]
[11431, 11444]
[11445, 11451]
[11454, 11458]
[11460, 11462]
[11463, 11464]
[11467, 11469]
[11476, 11480]
[11485, 11490]
[11494, 11503]
[11508, 11513]
[11514, 11522]
[11525, 11526]
[11533, 11534]
[11538, 11542]
[11554, 11561]
[11563, 11564]
[11570, 11571]
[11573, 11574]
[11577, 11587]
[11588, 11603]
[11604, 11606]
[11607, 11608]
[11624, 11625]
[11626, 11627]
[11633, 11644]
[11649, 11651]
[11653, 11657]
[11660, 11664]
[11668, 11669]
[11670, 11

[18841, 18843]
[18844, 18856]
[18863, 18864]
[18871, 18873]
[18877, 18878]
[18882, 18883]
[18888, 18894]
[18895, 18897]
[18899, 18901]
[18902, 18909]
[18918, 18919]
[18920, 18924]
[18925, 18928]
[18929, 18930]
[18933, 18937]
[18949, 18952]
[18954, 18956]
[18957, 18959]
[18961, 18962]
[18972, 18976]
[18977, 18980]
[18982, 18991]
[18992, 18993]
[19000, 19002]
[19004, 19005]
[19008, 19012]
[19014, 19017]
[19019, 19021]
[19031, 19032]
[19033, 19035]
[19040, 19041]
[19045, 19049]
[19051, 19052]
[19054, 19062]
[19064, 19066]
[19067, 19069]
[19072, 19077]
[19089, 19090]
[19093, 19094]
[19098, 19099]
[19100, 19102]
[19113, 19116]
[19117, 19118]
[19123, 19126]
[19127, 19134]
[19144, 19145]
[19146, 19149]
[19163, 19166]
[19169, 19173]
[19174, 19179]
[19183, 19186]
[19189, 19192]
[19201, 19206]
[19209, 19211]
[19212, 19215]
[19221, 19224]
[19226, 19227]
[19238, 19240]
[19241, 19244]
[19248, 19261]
[19262, 19268]
[19272, 19276]
[19284, 19286]
[19287, 19294]
[19296, 19299]
[19302, 19303]
[19305, 19

[26838, 26840]
[26846, 26848]
[26850, 26857]
[26862, 26866]
[26868, 26869]
[26883, 26884]
[26885, 26888]
[26889, 26892]
[26893, 26898]
[26902, 26903]
[26907, 26915]
[26916, 26925]
[26926, 26927]
[26928, 26931]
[26940, 26943]
[26946, 26950]
[26953, 26965]
[26969, 26974]
[26975, 26981]
[26998, 27001]
[27003, 27016]
[27017, 27024]
[27028, 27031]
[27032, 27033]
[27037, 27043]
[27046, 27048]
[27051, 27052]
[27053, 27057]
[27061, 27064]
[27065, 27067]
[27069, 27070]
[27083, 27089]
[27092, 27101]
[27108, 27112]
[27113, 27117]
[27124, 27131]
[27145, 27147]
[27152, 27155]
[27160, 27163]
[27164, 27173]
[27175, 27177]
[27184, 27187]
[27191, 27196]
[27198, 27201]
[27208, 27209]
[27210, 27215]
[27217, 27218]
[27226, 27228]
[27230, 27232]
[27234, 27235]
[27237, 27240]
[27241, 27243]
[27246, 27248]
[27249, 27250]
[27251, 27255]
[27259, 27262]
[27267, 27274]
[27283, 27285]
[27288, 27290]
[27296, 27297]
[27302, 27304]
[27306, 27316]
[27319, 27321]
[27323, 27327]
[27331, 27339]
[27343, 27346]
[27348, 27

[34937, 34939]
[34943, 34949]
[34954, 34955]
[34957, 34958]
[34966, 34970]
[34977, 34979]
[34990, 34993]
[34994, 35000]
[35008, 35010]
[35014, 35018]
[35019, 35027]
[35033, 35038]
[35040, 35041]
[35042, 35043]
[35044, 35047]
[35053, 35059]
[35072, 35074]
[35078, 35081]
[35082, 35089]
[35092, 35094]
[35095, 35096]
[35098, 35102]
[35105, 35109]
[35110, 35111]
[35120, 35124]
[35128, 35129]
[35131, 35136]
[35137, 35140]
[35144, 35146]
[35150, 35153]
[35158, 35172]
[35176, 35178]
[35180, 35183]
[35184, 35193]
[35198, 35204]
[35215, 35227]
[35251, 35256]
[35258, 35268]
[35274, 35283]
[35284, 35287]
[35291, 35295]
[35297, 35300]
[35306, 35308]
[35312, 35324]
[35326, 35327]
[35328, 35331]
[35333, 35336]
[35337, 35350]
[35352, 35353]
[35356, 35358]
[35359, 35362]
[35363, 35367]
[35369, 35371]
[35383, 35385]
[35386, 35403]
[35412, 35417]
[35423, 35424]
[35425, 35430]
[35431, 35432]
[35433, 35434]
[35438, 35442]
[35444, 35450]
[35451, 35457]
[35459, 35464]
[35465, 35468]
[35470, 35475]
[35477, 35

[43023, 43026]
[43027, 43032]
[43033, 43043]
[43044, 43046]
[43048, 43051]
[43055, 43056]
[43057, 43067]
[43080, 43083]
[43085, 43091]
[43092, 43103]
[43105, 43112]
[43114, 43116]
[43123, 43125]
[43126, 43129]
[43131, 43133]
[43140, 43144]
[43147, 43150]
[43151, 43153]
[43157, 43158]
[43159, 43164]
[43165, 43171]
[43173, 43174]
[43177, 43181]
[43184, 43187]
[43188, 43190]
[43202, 43210]
[43216, 43217]
[43220, 43238]
[43246, 43249]
[43250, 43254]
[43255, 43261]
[43274, 43276]
[43281, 43283]
[43293, 43297]
[43299, 43302]
[43303, 43304]
[43308, 43312]
[43313, 43318]
[43325, 43327]
[43328, 43336]
[43339, 43341]
[43351, 43362]
[43366, 43368]
[43371, 43376]
[43381, 43382]
[43387, 43398]
[43400, 43407]
[43408, 43419]
[43424, 43426]
[43428, 43435]
[43437, 43438]
[43451, 43461]
[43469, 43470]
[43476, 43481]
[43484, 43494]
[43495, 43496]
[43500, 43507]
[43510, 43512]
[43514, 43519]
[43520, 43521]
[43525, 43531]
[43533, 43546]
[43548, 43549]
[43550, 43566]
[43570, 43572]
[43580, 43598]
[43600, 43

[50414, 50416]
[50417, 50422]
[50423, 50425]
[50428, 50431]
[50433, 50434]
[50435, 50446]
[50449, 50451]
[50452, 50453]
[50457, 50460]
[50468, 50472]
[50477, 50479]
[50480, 50481]
[50487, 50493]
[50497, 50498]
[50500, 50504]
[50506, 50515]
[50522, 50527]
[50531, 50538]
[50539, 50545]
[50547, 50552]
[50553, 50557]
[50560, 50570]
[50577, 50585]
[50587, 50592]
[50599, 50603]
[50605, 50606]
[50612, 50614]
[50620, 50621]
[50634, 50635]
[50636, 50637]
[50640, 50650]
[50657, 50660]
[50661, 50662]
[50665, 50667]
[50671, 50678]
[50688, 50689]
[50693, 50696]
[50697, 50701]
[50705, 50709]
[50712, 50715]
[50716, 50722]
[50724, 50730]
[50733, 50735]
[50737, 50740]
[50743, 50744]
[50747, 50748]
[50766, 50770]
[50771, 50776]
[50788, 50800]
[50802, 50807]
[50808, 50809]
[50813, 50817]
[50821, 50830]
[50835, 50838]
[50841, 50843]
[50846, 50847]
[50852, 50853]
[50859, 50860]
[50862, 50863]
[50867, 50875]
[50884, 50892]
[50896, 50897]
[50913, 50916]
[50920, 50921]
[50922, 50923]
[50926, 50930]
[50934, 50

[55835, 55838]
[55841, 55843]
[55851, 55852]
[55855, 55858]
[55867, 55873]
[55876, 55885]
[55895, 55897]
[55898, 55901]
[55908, 55912]
[55917, 55918]
[55920, 55921]
[55926, 55927]
[55928, 55929]
[55931, 55932]
[55937, 55944]
[55945, 55954]
[55956, 55959]
[55963, 55972]
[55977, 55978]
[55979, 55981]
[55985, 55987]
[55993, 55994]
[55998, 56000]
[56001, 56003]
[56006, 56013]
[56015, 56016]
[56018, 56025]
[56027, 56029]
[56032, 56037]
[56039, 56043]
[56048, 56053]
[56067, 56076]
[56077, 56078]
[56085, 56086]
[56087, 56093]
[56096, 56099]
[56101, 56102]
[56107, 56108]
[56111, 56113]
[56114, 56116]
[56118, 56127]
[56131, 56132]
[56140, 56142]
[56155, 56161]
[56162, 56169]
[56170, 56191]
[56202, 56215]
[56220, 56222]
[56226, 56231]
[56232, 56234]
[56237, 56238]
[56242, 56249]
[56250, 56251]
[56252, 56256]
[56267, 56269]
[56270, 56279]
[56280, 56283]
[56284, 56285]
[56286, 56294]
[56306, 56311]
[56320, 56332]
[56334, 56343]
[56344, 56350]
[56354, 56358]
[56359, 56361]
[56362, 56365]
[56366, 56

[62753, 62757]
[62761, 62768]
[62770, 62771]
[62772, 62774]
[62775, 62776]
[62782, 62786]
[62789, 62794]
[62800, 62803]
[62815, 62823]
[62825, 62827]
[62829, 62833]
[62836, 62847]
[62849, 62850]
[62852, 62853]
[62855, 62856]
[62863, 62865]
[62866, 62868]
[62870, 62873]
[62884, 62888]
[62890, 62895]
[62901, 62911]
[62912, 62913]
[62923, 62929]
[62930, 62938]
[62943, 62944]
[62955, 62956]
[62959, 62962]
[62967, 62974]
[62978, 62979]
[62980, 62989]
[62994, 62995]
[62997, 62999]
[63006, 63007]
[63016, 63019]
[63023, 63029]
[63037, 63039]
[63053, 63057]
[63059, 63060]
[63062, 63067]
[63074, 63075]
[63077, 63079]
[63081, 63085]
[63087, 63097]
[63098, 63099]
[63105, 63108]
[63109, 63111]
[63114, 63115]
[63118, 63122]
[63124, 63125]
[63127, 63132]
[63146, 63148]
[63158, 63164]
[63165, 63167]
[63169, 63170]
[63174, 63178]
[63180, 63182]
[63194, 63199]
[63200, 63202]
[63205, 63207]
[63215, 63217]
[63219, 63236]
[63237, 63238]
[63239, 63241]
[63248, 63251]
[63252, 63254]
[63255, 63256]
[63262, 63

[283, 286]
[290, 291]
[294, 298]
[301, 303]
[305, 306]
[308, 316]
[317, 321]
[322, 329]
[333, 334]
[336, 337]
[346, 357]
[358, 361]
[362, 364]
[373, 379]
[382, 383]
[385, 390]
[391, 404]
[412, 416]
[422, 425]
[427, 430]
[433, 436]
[437, 440]
[442, 444]
[447, 453]
[471, 473]
[475, 477]
[478, 483]
[487, 488]
[491, 492]
[494, 504]
[506, 507]
[508, 509]
[514, 525]
[547, 549]
[555, 556]
[557, 560]
[567, 573]
[582, 583]
[594, 604]
[608, 611]
[621, 623]
[625, 628]
[638, 643]
[645, 648]
[654, 656]
[660, 661]
[662, 668]
[676, 679]
[683, 691]
[694, 695]
[708, 709]
[723, 725]
[726, 728]
[729, 736]
[737, 743]
[746, 750]
[755, 756]
[757, 760]
[766, 769]
[784, 785]
[789, 790]
[797, 798]
[799, 802]
[803, 806]
[807, 809]
[824, 826]
[832, 835]
[838, 840]
[841, 842]
[844, 847]
[850, 855]
[859, 861]
[863, 865]
[869, 870]
[876, 878]
[882, 884]
[887, 892]
[897, 898]
[903, 907]
[908, 909]
[911, 914]
[919, 920]
[921, 924]
[927, 935]
[936, 943]
[946, 949]
[960, 963]
[968, 971]
[972, 974]
[977, 982]
[988, 995]

[6524, 6532]
[6542, 6543]
[6548, 6550]
[6554, 6555]
[6558, 6562]
[6564, 6566]
[6569, 6572]
[6581, 6583]
[6585, 6598]
[6601, 6607]
[6611, 6614]
[6618, 6619]
[6624, 6627]
[6629, 6631]
[6650, 6654]
[6657, 6659]
[6660, 6661]
[6665, 6669]
[6671, 6674]
[6679, 6689]
[6691, 6692]
[6695, 6699]
[6702, 6711]
[6714, 6716]
[6719, 6721]
[6730, 6733]
[6734, 6735]
[6736, 6744]
[6747, 6751]
[6752, 6758]
[6768, 6770]
[6771, 6776]
[6780, 6781]
[6783, 6784]
[6786, 6791]
[6792, 6794]
[6796, 6800]
[6809, 6810]
[6815, 6820]
[6824, 6826]
[6828, 6832]
[6833, 6842]
[6843, 6844]
[6857, 6858]
[6860, 6864]
[6865, 6869]
[6870, 6871]
[6872, 6875]
[6877, 6879]
[6881, 6887]
[6892, 6893]
[6904, 6906]
[6913, 6914]
[6916, 6917]
[6922, 6923]
[6928, 6931]
[6936, 6939]
[6940, 6942]
[6944, 6946]
[6947, 6950]
[6953, 6959]
[6963, 6974]
[6978, 6980]
[6983, 6993]
[6998, 7013]
[7022, 7028]
[7032, 7036]
[7037, 7038]
[7043, 7044]
[7054, 7062]
[7069, 7071]
[7085, 7087]
[7092, 7094]
[7131, 7132]
[7135, 7136]
[7139, 7144]
[7147, 7149]

[11590, 11596]
[11605, 11610]
[11612, 11613]
[11621, 11622]
[11628, 11629]
[11632, 11635]
[11636, 11637]
[11638, 11639]
[11642, 11643]
[11652, 11654]
[11658, 11661]
[11667, 11675]
[11683, 11684]
[11692, 11694]
[11699, 11702]
[11706, 11712]
[11717, 11720]
[11725, 11726]
[11729, 11733]
[11741, 11742]
[11743, 11749]
[11751, 11752]
[11756, 11758]
[11763, 11764]
[11767, 11769]
[11770, 11776]
[11777, 11782]
[11789, 11792]
[11795, 11798]
[11800, 11802]
[11806, 11809]
[11816, 11820]
[11824, 11831]
[11832, 11838]
[11839, 11841]
[11851, 11854]
[11856, 11878]
[11880, 11884]
[11887, 11894]
[11896, 11898]
[11904, 11919]
[11921, 11924]
[11933, 11939]
[11940, 11942]
[11951, 11953]
[11958, 11965]
[11974, 11977]
[11981, 11985]
[12001, 12006]
[12010, 12011]
[12021, 12027]
[12032, 12041]
[12042, 12045]
[12051, 12053]
[12069, 12073]
[12080, 12085]
[12087, 12108]
[12109, 12115]
[12117, 12118]
[12119, 12122]
[12123, 12125]
[12126, 12127]
[12131, 12137]
[12138, 12140]
[12145, 12158]
[12161, 12162]
[12164, 12

[21091, 21096]
[21098, 21100]
[21103, 21104]
[21106, 21109]
[21110, 21112]
[21115, 21118]
[21119, 21124]
[21129, 21131]
[21132, 21133]
[21134, 21138]
[21150, 21151]
[21153, 21163]
[21168, 21171]
[21172, 21181]
[21190, 21192]
[21194, 21201]
[21214, 21217]
[21221, 21222]
[21235, 21240]
[21241, 21242]
[21244, 21247]
[21251, 21252]
[21257, 21262]
[21267, 21269]
[21271, 21277]
[21280, 21288]
[21294, 21299]
[21300, 21309]
[21313, 21317]
[21318, 21319]
[21320, 21333]
[21338, 21339]
[21342, 21347]
[21348, 21352]
[21361, 21365]
[21367, 21371]
[21372, 21374]
[21380, 21382]
[21385, 21386]
[21395, 21399]
[21401, 21403]
[21404, 21406]
[21409, 21412]
[21419, 21424]
[21427, 21429]
[21439, 21442]
[21449, 21454]
[21457, 21460]
[21461, 21469]
[21471, 21472]
[21473, 21475]
[21484, 21487]
[21490, 21491]
[21497, 21503]
[21504, 21507]
[21508, 21511]
[21521, 21523]
[21530, 21532]
[21540, 21544]
[21550, 21551]
[21553, 21557]
[21560, 21562]
[21565, 21571]
[21573, 21575]
[21576, 21577]
[21582, 21586]
[21592, 21

[26338, 26339]
[26345, 26348]
[26349, 26354]
[26363, 26370]
[26384, 26386]
[26387, 26389]
[26392, 26400]
[26403, 26412]
[26413, 26415]
[26426, 26431]
[26433, 26435]
[26447, 26451]
[26454, 26460]
[26463, 26479]
[26486, 26488]
[26490, 26491]
[26494, 26513]
[26519, 26521]
[26525, 26533]
[26535, 26544]
[26546, 26547]
[26550, 26551]
[26553, 26558]
[26560, 26565]
[26566, 26567]
[26572, 26573]
[26574, 26577]
[26585, 26586]
[26602, 26608]
[26610, 26615]
[26616, 26629]
[26630, 26631]
[26641, 26651]
[26654, 26657]
[26675, 26680]
[26686, 26688]
[26689, 26692]
[26696, 26697]
[26698, 26702]
[26710, 26711]
[26714, 26715]
[26721, 26722]
[26725, 26732]
[26735, 26736]
[26739, 26741]
[26744, 26745]
[26747, 26763]
[26765, 26767]
[26780, 26782]
[26785, 26796]
[26801, 26805]
[26807, 26810]
[26811, 26814]
[26816, 26820]
[26821, 26822]
[26827, 26831]
[26832, 26841]
[26842, 26844]
[26852, 26855]
[26858, 26860]
[26865, 26867]
[26870, 26873]
[26874, 26875]
[26882, 26905]
[26911, 26914]
[26921, 26922]
[26934, 26

[34385, 34388]
[34389, 34391]
[34392, 34393]
[34394, 34396]
[34402, 34404]
[34410, 34421]
[34422, 34430]
[34433, 34443]
[34448, 34449]
[34454, 34455]
[34457, 34459]
[34460, 34463]
[34467, 34468]
[34469, 34470]
[34474, 34482]
[34484, 34489]
[34496, 34498]
[34510, 34512]
[34518, 34522]
[34523, 34524]
[34532, 34534]
[34549, 34559]
[34561, 34567]
[34578, 34588]
[34589, 34590]
[34595, 34598]
[34599, 34603]
[34610, 34613]
[34626, 34627]
[34637, 34645]
[34647, 34650]
[34651, 34652]
[34657, 34660]
[34665, 34670]
[34680, 34688]
[34690, 34694]
[34699, 34702]
[34706, 34708]
[34712, 34717]
[34720, 34721]
[34726, 34729]
[34737, 34746]
[34750, 34751]
[34754, 34766]
[34767, 34770]
[34772, 34775]
[34776, 34780]
[34784, 34790]
[34791, 34792]
[34795, 34801]
[34804, 34807]
[34810, 34816]
[34819, 34820]
[34821, 34825]
[34826, 34833]
[34843, 34847]
[34855, 34861]
[34862, 34867]
[34870, 34876]
[34877, 34880]
[34884, 34887]
[34890, 34898]
[34902, 34904]
[34910, 34914]
[34922, 34928]
[34936, 34942]
[34946, 34

[42412, 42417]
[42419, 42421]
[42423, 42427]
[42439, 42442]
[42446, 42448]
[42449, 42450]
[42452, 42453]
[42456, 42462]
[42465, 42467]
[42469, 42470]
[42478, 42480]
[42481, 42483]
[42488, 42493]
[42497, 42498]
[42502, 42503]
[42510, 42511]
[42515, 42516]
[42521, 42525]
[42530, 42535]
[42541, 42542]
[42544, 42549]
[42550, 42558]
[42563, 42565]
[42570, 42573]
[42579, 42583]
[42587, 42594]
[42595, 42597]
[42611, 42615]
[42616, 42625]
[42626, 42629]
[42630, 42639]
[42640, 42643]
[42644, 42646]
[42648, 42649]
[42651, 42654]
[42656, 42662]
[42666, 42668]
[42671, 42681]
[42683, 42685]
[42689, 42691]
[42692, 42697]
[42705, 42709]
[42710, 42717]
[42720, 42728]
[42729, 42731]
[42732, 42733]
[42736, 42737]
[42741, 42742]
[42745, 42748]
[42755, 42756]
[42759, 42761]
[42765, 42768]
[42773, 42774]
[42783, 42786]
[42787, 42791]
[42794, 42799]
[42800, 42805]
[42813, 42814]
[42816, 42818]
[42827, 42830]
[42837, 42840]
[42842, 42848]
[42857, 42861]
[42869, 42872]
[42874, 42875]
[42876, 42877]
[42881, 42

[48499, 48500]
[48501, 48508]
[48520, 48523]
[48526, 48539]
[48540, 48550]
[48551, 48552]
[48554, 48559]
[48569, 48575]
[48576, 48578]
[48581, 48582]
[48584, 48588]
[48592, 48598]
[48605, 48607]
[48609, 48619]
[48624, 48625]
[48628, 48634]
[48640, 48641]
[48642, 48644]
[48646, 48651]
[48654, 48655]
[48656, 48657]
[48660, 48663]
[48668, 48669]
[48670, 48671]
[48672, 48674]
[48680, 48687]
[48688, 48694]
[48700, 48710]
[48714, 48715]
[48722, 48726]
[48729, 48730]
[48745, 48747]
[48749, 48753]
[48755, 48757]
[48766, 48768]
[48771, 48772]
[48776, 48777]
[48778, 48781]
[48786, 48788]
[48790, 48796]
[48800, 48801]
[48804, 48808]
[48809, 48813]
[48815, 48828]
[48830, 48837]
[48848, 48852]
[48853, 48856]
[48857, 48859]
[48861, 48863]
[48864, 48867]
[48870, 48872]
[48878, 48881]
[48882, 48887]
[48888, 48889]
[48898, 48904]
[48906, 48911]
[48929, 48930]
[48938, 48941]
[48942, 48951]
[48957, 48959]
[48961, 48971]
[48972, 48973]
[48974, 48976]
[48980, 48985]
[48988, 48996]
[49017, 49018]
[49019, 49

[56223, 56230]
[56233, 56235]
[56239, 56246]
[56248, 56253]
[56255, 56258]
[56263, 56266]
[56268, 56273]
[56276, 56289]
[56291, 56292]
[56295, 56298]
[56299, 56301]
[56309, 56310]
[56313, 56318]
[56321, 56324]
[56330, 56331]
[56335, 56337]
[56347, 56353]
[56355, 56363]
[56374, 56376]
[56378, 56385]
[56391, 56394]
[56397, 56399]
[56402, 56404]
[56414, 56417]
[56418, 56424]
[56427, 56428]
[56431, 56433]
[56434, 56439]
[56444, 56445]
[56448, 56449]
[56453, 56460]
[56461, 56462]
[56463, 56466]
[56467, 56469]
[56472, 56475]
[56481, 56482]
[56488, 56489]
[56493, 56500]
[56508, 56509]
[56513, 56525]
[56526, 56538]
[56545, 56547]
[56549, 56554]
[56558, 56559]
[56564, 56570]
[56575, 56579]
[56581, 56582]
[56584, 56589]
[56598, 56604]
[56605, 56606]
[56609, 56614]
[56618, 56620]
[56625, 56632]
[56637, 56638]
[56643, 56651]
[56655, 56656]
[56658, 56661]
[56662, 56663]
[56674, 56676]
[56680, 56682]
[56686, 56688]
[56697, 56698]
[56699, 56700]
[56709, 56726]
[56728, 56735]
[56743, 56744]
[56747, 56

[63227, 63230]
[63231, 63240]
[63242, 63245]
[63250, 63253]
[63260, 63273]
[63275, 63283]
[63284, 63287]
[63288, 63294]
[63297, 63304]
[63309, 63314]
[63316, 63319]
[63320, 63329]
[63332, 63337]
[63339, 63343]
[63344, 63345]
[63350, 63353]
[63354, 63361]
[63371, 63374]
[63376, 63379]
[63382, 63386]
[63387, 63388]
[63398, 63400]
[63404, 63405]
[63411, 63413]
[63420, 63424]
[63426, 63430]
[63431, 63433]
[63440, 63446]
[63447, 63452]
[63456, 63464]
[63466, 63467]
[63470, 63471]
[63472, 63482]
[63497, 63500]
[63520, 63521]
[63527, 63528]
[63535, 63537]
[63545, 63550]
[63552, 63554]
[63562, 63565]
[63568, 63569]
[63570, 63574]
[63579, 63580]
[63582, 63587]
[63588, 63589]
[63592, 63600]
[63603, 63620]
[63628, 63631]
[63638, 63644]
[63652, 63654]
[63657, 63658]
[63660, 63676]
[63680, 63682]
[63697, 63705]
[63711, 63712]
[63715, 63720]
[63729, 63732]
[63735, 63736]
[63739, 63742]
[63744, 63745]
[63759, 63766]
[63775, 63776]
[63779, 63780]
[63787, 63789]
[63798, 63800]
[63804, 63806]
[63807, 63

### Задача 5
Рон — счастливый обладатель перестановки 𝑎 длины 𝑛.

Перестановкой длины 𝑛 является массив, состоящий из 𝑛 различных целых чисел от 1 до 𝑛 в произвольном порядке. Например, [2,3,1,5,4] — перестановка, но [1,2,2] — не перестановка (2 встречается в массиве дважды) и [1,3,4] тоже не перестановка (𝑛=3, но в массиве встречается 4).


Над перестановкой Рона ставится 𝑚 экспериментов следующего вида: (𝑟𝑖, 𝑝𝑖). Это обозначает, что элементы в диапазоне [1,𝑟𝑖] (другими словами, префикс длины 𝑟𝑖) будут отсортированы в возрастающем порядке с вероятностью 𝑝𝑖. Все эксперименты проводятся в том же порядке, в котором задаются во входных данных.

Для примера рассмотрим перестановку [4,2,1,5,3] и эксперимент (3,0.6). После такого эксперимента с вероятностью 60% перестановка примет вид [1,2,4,5,3], а с вероятностью 40% останется без изменений.

Вам требуется определить, с какой вероятностью перестановка станет полностью отсортированной по возрастанию после 𝑚 экспериментов.

Входные данные\
Каждый тест содержит один или несколько наборов входных данных. В первой строке записано количество наборов входных данных 𝑡 (1≤𝑡≤100).

Первая строка каждого набора входных данных содержит два целых числа 𝑛 и 𝑚 (1≤𝑛,𝑚≤105) — количество элементов в перестановке и количество экспериментов.

Вторая строка каждого набора входных данных содержит 𝑛 целых чисел 𝑎1,𝑎2,…,𝑎𝑛 (1≤𝑎𝑖≤𝑛) — описание перестановки.

Каждая из следующих 𝑚 входных данных содержат целое число 𝑟𝑖 и вещественное число 𝑝𝑖 (1≤𝑟𝑖≤𝑛,0≤𝑝𝑖≤1) — длина префикса массива и вероятность, с которой он отсортируется. Все вероятности даны с точностью не более 6 знаков после запятой.

Гарантируется, что сумма 𝑛 и сумма 𝑚 не превосходят 105 (∑𝑛,∑𝑚≤105).

Выходные данные\
Для каждого набора входных данных выведите единственное число — вероятность, что после всех экспериментов перестановка окажется отсортированной. Ваш ответ будет считаться правильным, если его абсолютная или относительная ошибка не превосходит 10−6.

Формально, пусть ваш ответ равен 𝑎, а ответ жюри равен 𝑏. Ваш ответ будет зачтен, если и только если |𝑎−𝑏|max(1,|𝑏|)≤10−6.

Пример\
входные данные\
4\
4 3\
4 3 2 1\
1 0.3\
3 1\
4 0.6\
5 3\
4 2 1 3 5\
3 0.8\
4 0.6\
5 0.3\
6 5\
1 3 2 4 5 6\
4 0.9\
5 0.3\
2 0.4\
6 0.7\
3 0.5\
4 2\
1 2 3 4\
2 0.5\
4 0.1\
выходные данные\
0.600000\
0.720000\
0.989500\
1.000000\
Примечание\
Для первого набора входных данных можно показать, что только от того, выполнится ли сортировка с помощью правила (4,0.6), зависит, будет ли итоговая перестановка отсортированной.

In [20]:
import random
from functools import reduce


def solution(arr, cort):
    not_sorted = 0
    p_arr = []
    for i in range(len(arr)):
        if arr[i] != i + 1:
            not_sorted = i
    if not_sorted:
        for i in range(len(cort)):
            if cort[i][0] > not_sorted:
                p_arr.append(cort[i][1])
        for i in range(len(p_arr)):
            p_arr[i] = 1 - p_arr[i]

        if p_arr:
            res = 1 - reduce(lambda a, b: a * b, p_arr)
        else:
            res = 0
    else:
        res = 1
    return res


print("Ввод t (1<=t<=100):", end="")
t = int(input())
for j in range(t):
    n = random.randint(1, 105)
    m = 105 - n
    print(n, m)
    arr = []
    cort = []
    for i in range(n):
        arr.append(random.randint(1,n))
    print(arr)
    for i in range(m):
        cort.append((random.randint(1,n), random.randint(0,10)/10))
    for i in cort:
        print(i)
    print(f"{solution(arr, cort):.{6}f}")
#Проверки алгоритма:
'''arr = [4, 3, 2, 1]
cort = [(1, 0.3), (3, 1), (4, 0.6)]
print(f"{solution(arr, cort):.{6}f}")'''
'''arr = [4, 2, 1, 3, 5]
cort = [(3, 0.8), (4, 0.6), (5, 0.3)]
print(f"{solution(arr, cort):.{6}f}")'''
'''arr = [1, 2, 3, 4]
cort = [(2, 0.5), (4, 0.1)]
print(f"{solution(arr, cort):.{6}f}")'''
'''arr = [1, 3, 2, 4, 5, 6]
cort = [(4, 0.9), (5, 0.3), (2, 0.4), (6, 0.7), (3, 0.5)]
print(f"{solution(arr, cort):.{6}f}")'''

Ввод t (1<=t<=100):3
10 95
[6, 3, 6, 7, 7, 4, 10, 7, 8, 3]
(3, 0.6)
(2, 0.0)
(7, 0.8)
(1, 0.1)
(9, 0.5)
(2, 0.4)
(3, 0.2)
(1, 0.1)
(7, 0.6)
(10, 0.8)
(8, 0.6)
(10, 0.4)
(7, 0.0)
(7, 0.0)
(10, 0.0)
(7, 0.5)
(3, 0.8)
(3, 0.3)
(10, 0.2)
(4, 0.5)
(3, 0.7)
(2, 0.0)
(2, 0.7)
(9, 0.7)
(10, 0.1)
(4, 0.4)
(6, 0.3)
(2, 0.6)
(1, 0.2)
(9, 0.9)
(7, 0.0)
(3, 0.9)
(3, 0.8)
(6, 1.0)
(8, 0.4)
(1, 0.4)
(4, 0.6)
(7, 0.6)
(2, 1.0)
(1, 0.5)
(3, 0.3)
(6, 0.2)
(1, 0.8)
(6, 0.4)
(4, 0.3)
(3, 0.0)
(7, 0.5)
(2, 0.4)
(9, 1.0)
(1, 0.1)
(10, 0.1)
(3, 0.5)
(7, 0.5)
(7, 0.2)
(7, 0.9)
(3, 0.4)
(9, 0.3)
(8, 1.0)
(1, 0.9)
(4, 0.9)
(9, 1.0)
(5, 0.8)
(6, 0.4)
(10, 0.8)
(10, 0.5)
(2, 0.9)
(6, 1.0)
(3, 0.0)
(4, 0.6)
(9, 0.2)
(4, 0.8)
(10, 0.5)
(10, 0.5)
(8, 0.8)
(4, 1.0)
(7, 1.0)
(3, 1.0)
(3, 0.6)
(6, 0.6)
(9, 0.6)
(1, 0.2)
(10, 0.0)
(6, 0.3)
(9, 1.0)
(4, 0.9)
(8, 0.0)
(9, 1.0)
(6, 0.6)
(2, 0.5)
(4, 0.4)
(5, 0.2)
(7, 0.3)
(5, 0.5)
(8, 0.6)
(7, 0.7)
0.998056
32 73
[24, 26, 1, 27, 6, 5, 25, 21, 28, 25, 11, 17, 13, 21, 1, 21,

'arr = [1, 3, 2, 4, 5, 6]\ncort = [(4, 0.9), (5, 0.3), (2, 0.4), (6, 0.7), (3, 0.5)]\nprint(f"{solution(arr, cort):.{6}f}")'

### Задача 6
Вот и подошел к концу Beautiful Regional Contest (BeRC)! В соревновании приняли участие 𝑛 студентов. Уже известна таблица финальных результатов — участник на 𝑖-м месте решил 𝑝𝑖 задач. Так как в первую очередь участники упорядочиваются по количеству решенных задач, то 𝑝1≥𝑝2≥⋯≥𝑝𝑛.

Помогите жюри распределить золотые, серебряные и бронзовые медали. Пусть их количества равны 𝑔, 𝑠 и 𝑏, соответственно. Вот перечень требований из регламента, которые все должны быть удовлетворены:

для каждого из трёх достоинств медалей хотя бы одна медаль должна быть вручена (то есть 𝑔>0, 𝑠>0 и 𝑏>0);
количество золотых медалей должно быть строго меньше и количества серебряных и количества бронзовых (то есть 𝑔<𝑠 и 𝑔<𝑏, но требований между 𝑠 и 𝑏 нет);
каждый золотой медалист должен решить строго больше задач, чем любой награждённый серебряной медалью;
каждый серебряный медалист должен решить строго больше задач, чем любой награждённый бронзовой медалью;
каждый бронзовый медалист должен решить строго больше задач, чем любой не награждённый медалью участник;
суммарное количество медалистов 𝑔+𝑠+𝑏 не должно превышать половину всех участников (например, если 𝑛=21, то можно наградить максимум 10 участников, а если 𝑛=26, то можно наградить максимум 13 участников).
Жюри хочет наградить медалями суммарно наибольшее количество 𝑔+𝑠+𝑏 участников так, чтобы все перечисленные пункты выполнялись. Помогите жюри найти такой способ награждения медалями.

Входные данные\
В первой строке записано целое число 𝑡 (1≤𝑡≤10000) — количество наборов входных данных в тесте. Далее следуют 𝑡 наборов входных данных.

Первая строка набора входных данных содержит целое число 𝑛 (1≤𝑛≤4⋅105) — количество участников BeRC. Вторая строка набора содержит последовательность целых чисел 𝑝1,𝑝2,…,𝑝𝑛 (0≤𝑝𝑖≤106), где 𝑝𝑖 равно количеству задач, решенных 𝑖-м участником из таблицы финальных результатов. Значения 𝑝𝑖 отсортированы по невозрастанию, то есть 𝑝1≥𝑝2≥⋯≥𝑝𝑛.

Сумма значений 𝑛 по всем наборам входных данных в тесте не превосходит 4⋅105.

Выходные данные\
Выведите 𝑡 строк, 𝑗-я строка должна содержать ответ на 𝑗-й набор входных данных.

Ответ состоит из трёх неотрицательных целых чисел 𝑔,𝑠,𝑏.

Выведите 𝑔=𝑠=𝑏=0, если не существует способа наградить участников медалями, чтобы одновременно выполнялись все требования из условия.
В противном случае выведите три положительных числа 𝑔,𝑠,𝑏 — возможное количество золотых, серебряных и бронзовых медалей, соответственно. Сумма 𝑔+𝑠+𝑏 должна быть максимально возможной. Если ответов несколько, то выведите любой из них.
Пример\
входные данные\
5\
12\
5 4 4 3 2 2 1 1 1 1 1 1\
4\
4 3 2 1\
1\
1000000\
20\
20 19 18 17 16 15 14 13 12 11 10 9 8 7 6 5 4 3 2 1\
32\
64 64 63 58 58 58 58 58 37 37 37 37 34 34 28 28 28 28 28 28 24 24 19 17 17 17 17 16 16 16 16 11\
выходные данные\
1 2 3\
0 0 0\
0 0 0\
2 5 3\
2 6 6\
Примечание\
В первом тестовом случае можно наградить 1 золотой, 2 серебряными и 3 бронзовыми медалями участников. Тогда золотую медаль получит участник, решившая 5 задач, серебрянные медали получат участники, решившие 4 задачи, бронзовые медали получат участники решившие 2 или 3 задачи. Участники, решившие 1 задачу медаль не получат. Заметим, что все условия выполнены и раздать медали таким способом можно. Больше, чем 6 медалей выдать невозможно, потому что нельзя выдать больше чем половину от количества участников медалей. Ответ 1, 3, 2 также является верным в этом тестовом случае.

Во втором и третьем тестовых случаев нельзя так раздать медали, потому что медаль каждого достоинства должна получить хотя бы один участник, но количество выданных медалей не должно превышать половину от количества участников.

In [21]:
import random


def solution(n, p):
    medals = [0, 0, 0]
    amount = n // 2
    i = 0
    g_p = p[i]
    s_p = -1
    b_p = -1
    while medals[0] + medals[1] + medals[2] < amount and i < amount:
        if p[i] == g_p:
            medals[0] += 1
        else:
            if s_p == -1:
                s_p = p[i]
                medals[1] += 1
            elif p[i] == s_p or medals[0] >= medals[1]:
                medals[1] += 1
            else:
                if b_p == -1:
                    b_p = p[i]
                    medals[2] += 1
                elif p[i] <= b_p:
                    medals[2] += 1
        i += 1
    if p[i] == g_p or p[i] == s_p or p[i] == b_p or not(medals[0] < medals[1] and medals[0] < medals[2]):
        medals = [0, 0, 0]
    return medals

print("Ввод t (1<=t<=100):", end="")
t = int(input())
for i in range(t):

    n = random.randint(1,4*105)
    print(n)
    p = []
    for i in range(n):
        p.append(random.randint(0, 106))
    p.sort(reverse=True)
    print(p)
    print("Медали: ", solution(len(p), p))
#Проверки алгоритма:
'''p = [5, 4, 4, 3, 2, 2, 1, 1, 1, 1, 1, 1]
print(solution(len(p), p))'''
'''p = [4, 3, 2, 1]
print(solution(len(p), p))'''
'''p = [1000000]
print(solution(len(p), p))'''
'''p = [20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
print(solution(len(p), p))'''
'''p = [64, 64, 63, 58, 58, 58, 58, 58, 37, 37, 37, 37, 34, 34, 28, 28, 28, 28, 28, 28, 24, 24, 19, 17, 17, 17, 17, 16, 16, 16, 16, 11]
print(solution(len(p), p))'''

Ввод t (1<=t<=100):3
51
[106, 104, 102, 101, 100, 100, 98, 96, 93, 93, 91, 91, 88, 88, 83, 83, 83, 81, 79, 78, 77, 76, 76, 73, 71, 71, 71, 68, 65, 63, 62, 60, 58, 58, 56, 49, 35, 33, 32, 32, 27, 25, 20, 17, 15, 15, 13, 10, 10, 5, 3]
Медали:  [1, 2, 22]
128
[106, 106, 106, 104, 102, 96, 96, 96, 95, 95, 93, 92, 92, 92, 91, 88, 85, 84, 82, 82, 81, 80, 80, 79, 79, 78, 77, 76, 75, 75, 75, 73, 72, 71, 70, 70, 67, 67, 67, 67, 67, 66, 66, 65, 64, 64, 63, 63, 63, 61, 61, 60, 59, 59, 58, 58, 58, 57, 56, 56, 55, 55, 53, 52, 50, 50, 49, 47, 46, 45, 45, 44, 44, 40, 40, 38, 38, 37, 37, 37, 36, 34, 32, 32, 31, 31, 29, 29, 28, 28, 27, 27, 26, 25, 25, 24, 24, 23, 22, 21, 21, 20, 19, 18, 18, 18, 17, 15, 15, 12, 11, 11, 8, 8, 7, 7, 6, 5, 4, 4, 3, 1, 1, 1, 1, 0, 0, 0]
Медали:  [3, 4, 57]
77
[105, 105, 104, 103, 103, 102, 100, 100, 97, 97, 96, 94, 92, 88, 87, 85, 85, 84, 83, 83, 81, 81, 81, 80, 79, 74, 74, 73, 72, 72, 71, 70, 70, 69, 67, 65, 63, 62, 62, 60, 60, 60, 59, 58, 58, 57, 56, 56, 54, 53, 49, 48, 4

'p = [64, 64, 63, 58, 58, 58, 58, 58, 37, 37, 37, 37, 34, 34, 28, 28, 28, 28, 28, 28, 24, 24, 19, 17, 17, 17, 17, 16, 16, 16, 16, 11]\nprint(solution(len(p), p))'

### Задача 7
𝑛 человек собрались, чтобы провести заседание жюри предстоящего соревнования, 𝑖-й член жюри придумал 𝑎𝑖 задач, которыми он хочет поделиться с другими.

Сначала жюри выбирает порядок, которому они будут следовать при обсуждении задач. Пусть это будет перестановка 𝑝 чисел от 1 до 𝑛 (массив размера 𝑛, в котором каждое число от 1 до 𝑛 встречается ровно один раз).

Затем обсуждение происходит следующим образом:

если у члена жюри 𝑝1 остались задачи, которые нужно рассказать, то он рассказывает одну задачу. В противном случае он будет пропущен.
если у члена жюри 𝑝2 остались задачи, которые нужно рассказать, то он рассказывает одну задачу. В противном случае он будет пропущен.
...
если у члена жюри 𝑝𝑛 остались задачи, которые нужно рассказать, то он рассказывает одну задачу. В противном случае он будет пропущен.
если остались члены жюри, которые еще не рассказали все свои задачи, то процесс повторяется с самого начала. В противном случае обсуждение заканчивается./
Перестановка 𝑝 хорошая, если никто из членов жюри не будет рассказывать две или более задач подряд.

Подсчитайте количество хороших перестановок. Ответ может быть очень большим, поэтому выведите его по модулю 

Входные данные/
Первая строка содержит одно целое число 𝑡 (1≤𝑡≤104) — количество наборов входных данных.

Первая строка набора входных данных содержит одно целое число 𝑛 (2≤𝑛≤2⋅105) — количество членов жюри.

Вторая строка содержит 𝑛 целых чисел 𝑎1,𝑎2,…,𝑎𝑛 (1≤𝑎𝑖≤109) — количество задач, которые придумал 𝑖-й член жюри.

Сумма 𝑛 по всем наборам входных данных не превосходит 2⋅105.

Выходные данные\
Для каждого набора входных данных выведите одно целое число — количество хороших перестановок, взятое по модулю 998244353.

Пример\
входные данные\
4\
2\
1 2\
3\
5 5 5\
4\
1 3 3 7\
6\
3 4 2 1 3 3\
выходные данные\
1\
6\
0\
540\
Примечание\
Объяснение первого примера:

Существуют две возможные перестановки, 𝑝=[1,2] и 𝑝=[2,1]. Для 𝑝=[1,2] процесс происходит следующим образом:

первый член жюри рассказывает задачу;\
второй член жюри рассказывает задачу;\
у первого члена жюри не осталось задач, которые нужно было бы рассказать, поэтому он будет пропущен;
второй член жюри рассказывает задачу.
Второй член жюри рассказал две задачи подряд, поэтому перестановка не является хорошей.

Для 𝑝=[2,1] процесс происходит следующим образом:

второй член жюри рассказывает задачу;\
первый член жюри рассказывает задачу;\
второй член жюри рассказывает задачу.\
Эта перестановка является хорошей.

In [24]:
import random


def solution(arr, n):
    MOD = 998244353
    max_tasks = max(arr)
    indices = [i for i,x in enumerate(arr) if x == max_tasks]
    max_tasks_count = len(indices)
    indices = [i for i, x in enumerate(arr) if x == max_tasks - 1]
    tasks_count = len(indices)
    res = 1
    for i in range(1, n + 1):
        res = res * i % MOD
    if max_tasks_count == 1:
        res = (res - (res/(tasks_count + 1)))%MOD
    return int(res)


print("Ввод t (1<=t<=104):", end="")
t = int(input())
for i in range(t):
    n = random.randint(2, 2 * 105)
    print(n)
    arr = []
    for i in range(n):
        arr.append(random.randint(1, 109))
    print(arr)
    print(solution(arr, n))
#Проверки алгоритма:
'''n = 2
arr = [1, 2]
print(solution(arr, n))'''
'''n = 3
arr = [5, 5, 5]
print(solution(arr, n))'''
'''n = 4
arr = [1, 3, 3, 7]
print(solution(arr, n))'''
'''n = 6
arr = [3, 4, 2, 1, 3, 3]
print(solution(arr, n))'''

Ввод t (1<=t<=104):3
122
[37, 79, 83, 90, 4, 19, 106, 63, 93, 18, 28, 6, 50, 96, 42, 42, 15, 31, 32, 31, 63, 20, 82, 84, 103, 88, 34, 35, 89, 71, 1, 13, 102, 66, 107, 73, 9, 109, 26, 55, 23, 78, 57, 17, 86, 75, 55, 75, 97, 62, 24, 15, 109, 103, 8, 5, 82, 17, 55, 22, 29, 96, 82, 60, 100, 58, 89, 46, 46, 18, 90, 47, 69, 86, 49, 40, 94, 87, 32, 18, 60, 12, 45, 62, 83, 66, 28, 27, 35, 15, 89, 42, 43, 101, 14, 57, 72, 83, 96, 80, 39, 30, 43, 77, 56, 29, 32, 29, 1, 14, 70, 69, 86, 13, 69, 17, 19, 82, 18, 59, 26, 8]
385362667
113
[34, 104, 50, 31, 30, 28, 27, 54, 65, 98, 20, 62, 3, 96, 50, 101, 25, 7, 37, 50, 84, 5, 83, 1, 98, 76, 85, 2, 48, 34, 97, 11, 77, 66, 62, 2, 85, 96, 70, 102, 85, 65, 87, 94, 94, 91, 104, 58, 93, 27, 60, 53, 1, 73, 55, 77, 23, 26, 7, 6, 67, 48, 5, 54, 77, 71, 97, 10, 27, 61, 51, 91, 60, 82, 89, 44, 18, 81, 57, 92, 20, 57, 55, 7, 99, 50, 35, 28, 70, 105, 17, 85, 63, 108, 104, 109, 82, 2, 56, 93, 51, 95, 81, 84, 49, 51, 33, 66, 1, 89, 94, 90, 109]
275137452
16
[45, 54, 

'n = 6\narr = [3, 4, 2, 1, 3, 3]\nprint(solution(arr, n))'

### Задача 8
У Аюба был массив целых чисел 𝑎 длины 𝑛 и у этого массива было два необычных свойства:

Все целые числа в этом массиве были от 𝑙 до 𝑟 (включительно).
Сумма всех элементов в массиве делилась на 3.
К сожалению, Аюб потерял свой массив, но он помнит размер массива 𝑛, а также числа 𝑙 и 𝑟, так что он попросил вас посчитать количество способов восстановить массив по этим данным.

Так как ответ может быть очень большим, выведите его по модулю 109+7 (то есть остаток при делении на 109+7). В случае если не существует ни одного подходящего массива (Аюб что-то перепутал) — выведите 0.

Входные данные\
Первая и единственная строка содержит три целых числа 𝑛, 𝑙 и 𝑟 (1≤𝑛≤2⋅105,1≤𝑙≤𝑟≤109) — размер массива и диапазон чисел в нём.

Выходные данные\
Выведите остаток при делении на 109+7 количества способов восстановить массив.

Примеры\
входные данные\
2 1 3\
выходные данные\
3\
входные данные\
3 2 2\
выходные данные\
1\
входные данные\
9 9 99\
выходные данные\
711426616\
Примечание\
В первом примере возможны массивы [1,2],[2,1],[3,3].

Во втором примере единственный возможный массив — [2,2,2].

In [26]:
def solution(n, l, r):
    mod = 1e9+7
    num = [0]*3
    dp = []
    for i in range(n):
        dp.append([0, 0, 0])
    num[0] = r//3 - (l - 1)//3
    num[1] = num[0] - 1
    if l % 3 == 1:
        num[1] += 1
    if r % 3 == 1 or r % 3 == 2:
        num[1] += 1
    num[2] = num[0] - 1
    if l % 3 == 1 or l % 3 == 2:
        num[2] += 1
    if r % 3 == 2:
        num[2] += 1
    dp[0][0] = num[0]
    dp[0][1] = num[1]
    dp[0][2] = num[2]
    for i in range(1, n):
        dp[i][0] = ((dp[i - 1][0] * num[0]) % mod + (dp[i - 1][1] * num[2]) % mod + (dp[i - 1][2] * num[1]) % mod) % mod
        dp[i][1] = ((dp[i - 1][0] * num[1]) % mod + (dp[i - 1][1] * num[0]) % mod + (dp[i - 1][2] * num[2]) % mod) % mod
        dp[i][2] = ((dp[i - 1][0] * num[2]) % mod + (dp[i - 1][1] * num[1]) % mod + (dp[i - 1][2] * num[0]) % mod) % mod
    return int(dp[n-1][0])
#Проверки алгоритма:
n = 9
l = 9
r = 99
'''n = 2
l = 1
r = 3'''
'''n = 3
l = 2
r = 2'''
print(n, l, r)
print(solution(n, l, r))

9 9 99
711426616


### Задача 9
Задано два массива 𝑎 и 𝑏, каждый состоит из 𝑛 целых чисел.

Вы хотите создать новый массив 𝑐 следующим образом: выбрать какое-то вещественное (то есть не обязательно целое) число 𝑑, и для каждого 𝑖∈[1,𝑛] присвоить 𝑐𝑖:=𝑑⋅𝑎𝑖+𝑏𝑖.

Ваша задача — максимизировать количество нулей в массиве 𝑐. Чему равен максимально возможный ответ, если вы выберете значение 𝑑 оптимально?

Входные данные\
Первая строка входных данных содержит одно целое число 𝑛 (1≤𝑛≤2⋅105) — количество элементов в обоих массивах.

Вторая строка входных данных содержит 𝑛 целых чисел 𝑎1, 𝑎2, ..., 𝑎𝑛 (−109≤𝑎𝑖≤109).

Третья строка входных данных содержит 𝑛 целых чисел 𝑏1, 𝑏2, ..., 𝑏𝑛 (−109≤𝑏𝑖≤109).

Выходные данные\
Выведите одно целое число — максимальное количество нулей в массиве 𝑐, если вы выберете 𝑑 оптимально.

Примеры\
входные данные\
5\
1 2 3 4 5\
2 4 7 11 3\
выходные данные\
2\
входные данные\
3\
13 37 39\
1 2 3\
выходные данные\
2\
входные данные\
0 0 0 0\
1 2 3 4\
выходные данные\
0\
входные данные\
3\
1 2 -1\
-6 -12 6\
выходные данные\
3\
Примечание\
В первом тестовом примере мы можем выбрать 𝑑=−2.

Во втором тестовом примере мы можем выбрать 𝑑=−113.

В третьем тестовом примере мы не можем получить ни одного нуля в массиве 𝑐, какое бы значение 𝑑 мы ни выбрали.

В четвертом тестовом примере мы можем выбрать 𝑑=6.

In [21]:
import random


def solution(n, a, b):
    dict = {}
    arr = []
    for i in range(n):
        quot = (b[i]/a[i]) if a[i] else 0
        if quot in dict:
            dict[quot] += 1
        else:
            dict[quot] = 1
    d = 0 - sorted(dict.items(), key= lambda x : x[1]).pop()[0]
    for i in range(n):
        arr.append(d * a[i] + b[i])
    indices = [i for i, x in enumerate(arr) if x == 0]
    res = len(indices)
    return res


n = random.randint(1, 2*105)
print(n)
a = []
b = []
for i in range(n):
    a.append(random.randint(-109, 109))
    b.append(random.randint(-109, 109))
print(a)
print(b)
print(solution(n, a, b))
#Проверки алгоритма:
'''n = 5
a = [1, 2, 3, 4, 5]
b = [2, 4, 7, 11, 3]
print(solution(n, a, b))'''
'''n = 3
a = [13, 37, 39]
b = [1, 2, 3]
print(solution(n, a, b))'''
'''n = 4
a = [0, 0, 0, 0]
b = [1, 2, 3, 4]
print(solution(n, a, b))'''
'''n = 3
a = [1, 2, -1]
b = [-6, -12, 6]
print(solution(n, a, b))'''

7
[-22, 1, -7, 65, -9, 101, -44]
[10, 74, -5, -31, -10, -37, -33]
1


'n = 3\na = [1, 2, -1]\nb = [-6, -12, 6]\nprint(solution(n, a, b))'

### Задача 10
Число называется сильным, если оно является степенью двойки или факториалом. Другими словами, число 𝑚 сильное, если существует неотрицательное целое число 𝑑 такое, что 𝑚=2𝑑 или 𝑚=𝑑!, где 𝑑!=1⋅2⋅…⋅𝑑 (в частности, 0!=1). Например, 1, 4 и 6 являются сильными, потому что 1=1!, 4=22, 6=3!, а 7, 10 или 18 не являются.

Вам дано положительное число 𝑛. Найдите минимальное число 𝑘 такое, что 𝑛 можно представить в виде суммы 𝑘 различных сильных чисел, или определите, что такого 𝑘 не существует.

Входные данные\
Во входных данных находятся несколько наборов входных данных. В первой строке находится одно целое число 𝑡 (1≤𝑡≤100) — количество наборов входных данных. Далее следуют наборы входных данных.

Описание одного набора входных данных содержит одно целое число 𝑛 (1≤𝑛≤1012).

Выходные данные\
Для каждого набора входных данных выведите ответ на отдельной строке.

Если 𝑛 не может быть представлено как сумма различных сильных чисел, выведите −1.

Иначе выведите одно положительное число — минимально возможное число 𝑘.

Пример\
входные данные\
4\
7\
11\
240\
17179869184\
выходные данные\
2\
3\
4\
1\
Примечание\
В первом примере 7 может быть представлена как 7=1+6, где 1 и 6 являются сильными. Так как 7 не является сильным числом, мы знаем, что минимальное значение 𝑘 в этом случае 𝑘=2.

Во втором примере один из способов разложить 11 в сумму трех сильных чисел такой: 11=1+4+6. Можно показать, что нельзя представить 11 в виде суммы двух или менее сильных чисел.

В третьем примере 240 может быть представлено как 240=24+32+64+120. Обратите внимание, что 240=120+120 не является корректным представлением, т. к. сильные числа должны быть различны.

В четвертом примере 17179869184=234, поэтому 17179869184 является сильным числом, и минимальное 𝑘 в этом случае 𝑘=1.

In [23]:
import random


def solution(n):
    power_numbers = []
    nums = []
    sums = []
    for i in range(n):
        power = 2 ** i
        if power <= n:
            if power not in power_numbers:
                power_numbers.append(power)
        else:
            break
    for i in range(3, n):
        factorial = 1
        for j in range(1, i):
            factorial = factorial * j
        if factorial <= n:
            if factorial not in power_numbers:
                power_numbers.append(factorial)
        else:
            break
    power_numbers.sort(reverse=True)
    for i in range(len(power_numbers)):
        res = power_numbers[i]
        nums.append([power_numbers[i]])
        if res < n:
            for j in range(i + 1, len(power_numbers)):
                if res + power_numbers[j] < n:
                    res += power_numbers[j]
                    nums[i].append(power_numbers[j])
                elif res + power_numbers[j] == n:
                    nums[i].append(power_numbers[j])
                    sums.append((nums[i], len(nums[i])))
                    res = power_numbers[i]
                    nums[i] = [power_numbers[i]]
        elif res == n:
            sums.append((nums[i], len(nums[i])))
    if len(sums) == 0:
        return -1
    res = sorted(sums, key= lambda x : x[1], reverse=True).pop()[1]

    return res


print("Ввод t (1<=t<=100):", end="")
t = int(input())
for i in range(t):
    n = random.randint(1, 1012)
    print("n =", n)
    print("Ответ:", solution(n))
#Проверки алгоритма:
'''n = 7
print(solution(n))
n = 11
print(solution(n))
n = 240
print(solution(n))
n = 17179869184
print(solution(n))'''

Ввод t (1<=t<=100):3
n = 590
Ответ: 4
n = 211
Ответ: 5
n = 808
Ответ: 3


'n = 7\nprint(solution(n))\nn = 11\nprint(solution(n))\nn = 240\nprint(solution(n))\nn = 17179869184\nprint(solution(n))'